pkg/controller/azvolumeattachment.go

```diff
@@ -0,0 +1,546 @@
+			return reconcile.Result{Requeue: true}, err
+		}
+		// if the attachmentTier in status and spec are different, it is an update event where replica should be turned into a primary
+	} else if azVolumeAttachment.Spec.AttachmentTier != azVolumeAttachment.Status.AttachmentTier {
+		if azVolumeAttachment.Spec.AttachmentTier == v1alpha1.Primary {
+			// if primary exists for the specified volume, trigger detach for that primary
+			if primary, ok := r.primaryMap[azVolumeAttachment.Spec.UnderlyingVolume]; ok && primary.Name != azVolumeAttachment.Name {
+				if err := r.TriggerDetach(ctx, primary); err != nil {
+					return reconcile.Result{Requeue: true}, err
+				}
+			}
+			// update the replicaMap to avoid potential sync issue in `ManageReplicaForVolume` call
+			delete(r.replicaMap[azVolumeAttachment.Spec.UnderlyingVolume], azVolumeAttachment.Spec.NodeName)
+
+			if err := r.ManageReplicaForVolume(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+
+			// update the status's attachment tier to primary
+			updated := azVolumeAttachment.DeepCopy()
+			updated.Status = azVolumeAttachment.Status.DeepCopy()
+			updated.Status.AttachmentTier = v1alpha1.Primary
+			if _, err := r.azVolumeClient.DiskV1alpha1().AzVolumeAttachments(r.namespace).UpdateStatus(ctx, updated, metav1.UpdateOptions{}); err != nil {
+				klog.Errorf("failed to update status of azVolumeAttachment object: (%v)", err)
+				return reconcile.Result{Requeue: true}, err
+			}
+			r.primaryMap[azVolumeAttachment.Spec.UnderlyingVolume] = azVolumeAttachment
+		}
+	}
+	return reconcile.Result{Requeue: true}, nil
+}
+
+func (r *reconcileAzVolumeAttachment) HandleReplicaUpdate(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	storageClassName := request.Name
+	replicaCount, err := r.GetReplicaCount(ctx, storageClassName)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	for volume := range r.scMap[storageClassName] {
+		if int(replicaCount) != len(r.scMap[storageClassName]) {
+			err := r.ManageReplicaForVolume(ctx, volume)
+			if err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *reconcileAzVolumeAttachment) GetReplicaCount(ctx context.Context, name string) (uint64, error) {
+	var storageClass storagev1.StorageClass
+	if err := r.client.Get(ctx, types.NamespacedName{Name: name}, &storageClass); err != nil {
+		// suppress error message
+		return 0, err
+	}
+	var replicaCount uint64
+	var err error
+	if storageClass.Provisioner != DriverName {
+		return 0, errors.NewBadRequest(fmt.Sprintf("storageClass provisioner is required to be %s", DriverName))
+	}
+	// maxMountReplicaCount can either be explicitly defined or implicitly deduced from maxShares
+	if maxMountReplicaCount, ok := storageClass.Parameters["maxMountReplicaCount"]; ok {
```

Comment:
@abhisheksinghbaghel Should `maxMountReplicaCount` be explicitly bounded by `maxShares`? What should the behavior be if `maxMountReplicaCount` >= `maxShares`?

Reply:
yeah primary + replica mounts should be <= maxShares on the disk, so yeah it is bounded. If we try more attachments than what maxShares is set to then storage will start failing the attach requests.